src/fcmp_pp/blinds_cache.cpp

```diff
@@ -0,0 +1,520 @@
+        },
+        true);
+
+    return pending_branch_blind;
+}
+//----------------------------------------------------------------------------------------------------------------------
+template<>
+std::shared_ptr<BranchBlind> BlindsCache<Selene, Helios>::get_c2_branch_blind_async()
+{
+    std::shared_ptr<BranchBlind> pending_branch_blind = std::make_shared<BranchBlind>();
+
+    m_tpool->submit(&m_waiter,
+        [pending_branch_blind]()
+        {
+            // Do the expensive calculation
+            const uint8_t *branch_blind_ptr = fcmp_pp::helios_branch_blind();
+            *pending_branch_blind = fcmp_pp::helios_blind_to_buf(branch_blind_ptr);
+        },
+        true);
+
+    return pending_branch_blind;
+}
+//----------------------------------------------------------------------------------------------------------------------
+template<>
+void BlindsCache<Selene, Helios>::calc_needed_branch_blinds_async()
+{
+    std::lock_guard<std::mutex> guard(m_mutex);
+
+    while (true)
+    {
+        const std::size_t n_c1_branch_blinds = m_pending_c1_branch_blinds.size();
+        const std::size_t n_c2_branch_blinds = m_pending_c2_branch_blinds.size();
+
+        // Check if we need to calculate more branch blinds
+        if (need_more_c1_blinds(m_prepare_n_min_inputs, m_output_blindings.size(), m_n_tree_layers, n_c1_branch_blinds))
+        {
+            // Calculate c1 branch blind in the background
+            m_pending_c1_branch_blinds.push_back(this->get_c1_branch_blind_async());
+            continue;
+        }
+
+        if (need_more_c2_blinds(m_prepare_n_min_inputs, m_output_blindings.size(), m_n_tree_layers, n_c2_branch_blinds))
+        {
+            // Calculate c1 branch blind in the background
+            m_pending_c2_branch_blinds.push_back(this->get_c2_branch_blind_async());
+            continue;
+        }
+
+        break;
+    }
+}
+//----------------------------------------------------------------------------------------------------------------------
+template<typename C1, typename C2>
+void BlindsCache<C1, C2>::set_n_tree_layers(uint8_t n_tree_layers)
+{
+    bool increased = false;
+    {
+        std::lock_guard<std::mutex> guard(m_mutex);
+        increased = n_tree_layers > m_n_tree_layers;
+        m_n_tree_layers = n_tree_layers;
+    }
+
+    // If we increased n tree layers, we want to prepare more branch blinds
+    if (increased)
+        this->calc_needed_branch_blinds_async();
+}
+
+template void BlindsCache<Selene, Helios>::set_n_tree_layers(uint8_t n_tree_layers);
+//----------------------------------------------------------------------------------------------------------------------
+template<typename C1, typename C2>
+void BlindsCache<C1, C2>::add_output(const OutputPair &output)
+{
+    std::lock_guard<std::mutex> guard(m_mutex);
+
+    const auto output_ref = get_output_ref(output);
+    if (m_output_blindings.find(output_ref) != m_output_blindings.end())
```

Comment:
Since not reusing the iterator returned by `find`, for simplicity, you could write this as:
```suggestion
    if (m_output_blindings.count(output_ref))
```